AWSPinpoint/AWSPinpointEndpointProfile.m

```diff
@@ -335,20 +336,23 @@ - (NSDictionary *) toDictionary {
     return dictionary;
 }
 
++ (BOOL)supportsSecureCoding {
+    return YES;
+}
 
 - (id)initWithCoder:(NSCoder *)decoder {
     if (self = [super init]) {
-        _applicationId = [decoder decodeObjectForKey:@"applicationId"];
-        _endpointId = [decoder decodeObjectForKey:@"endpointId"];
-        _channelType = [decoder decodeObjectForKey:@"channelType"];
-        _address = [decoder decodeObjectForKey:@"address"];
-        _location = [decoder decodeObjectForKey:@"location"];
-        _demographic = [decoder decodeObjectForKey:@"demographic"];
-        _attributes = [decoder decodeObjectForKey:@"attributes"];
-        _metrics = [decoder decodeObjectForKey:@"metrics"];
-        _user = [decoder decodeObjectForKey:@"user"];
+        _applicationId = [decoder decodeObjectOfClass:[NSString class] forKey:@"applicationId"];
+        _endpointId = [decoder decodeObjectOfClass:[NSString class] forKey:@"endpointId"];
+        _channelType = [decoder decodeObjectOfClass:[NSString class] forKey:@"channelType"];
+        _address = [decoder decodeObjectOfClass:[NSString class] forKey:@"address"];
+        _optOutBackingVariable = [decoder decodeObjectOfClass:[NSString class] forKey:@"optOut"];
         _effectiveDate = [decoder decodeInt64ForKey:@"effectiveDate"];
-        _optOutBackingVariable = [decoder decodeObjectForKey:@"optOut"];
+        _location = [decoder decodeObjectOfClass:[AWSPinpointEndpointProfileLocation class] forKey:@"location"];
+        _demographic = [decoder decodeObjectOfClass:[AWSPinpointEndpointProfileDemographic class] forKey:@"demographic"];
+        _attributes = [decoder decodeObjectOfClass:[NSDictionary class] forKey:@"attributes"];
```

Comment:
Can we verify the contents of dictionary or array?

Reply:
Not to my knowledge. `NSDictionary` (or its mutable equivalent) is the actual class that gets instantiated, regardless of the type of the keys & values. It might be that a complex type (e.g., an `NSDictionary` with objects as keys) would require additional class declarations, but I haven't been able to verify that experimentally.